efibootmanager_policy.go

```diff
@@ -0,0 +1,335 @@
+
+	// 2) Initialize a hash algorithm context.
+	h := alg.NewHash()
+
+	// 3) Hash the image header from its base to immediately before the start of the checksum address in the optional header.
+	// This includes the DOS header, 4-byte PE signature, COFF header, and the first 64 bytes of the optional header.
+	b := make([]byte, int(coffHeaderOffset)+binary.Size(p.FileHeader)+64)
+	if _, err := hr.Read(b); err != nil {
+		return nil, xerrors.Errorf("cannot read from image to start to checksum: %w", err)
+	}
+	h.Write(b)
+
+	// 4) Skip over the checksum, which is a 4-byte field.
+	hr.Seek(4, io.SeekCurrent)
+
+	var certTable *pe.DataDirectory
+
+	if len(dd) > certTableIndex {
+		// 5) Hash everything from the end of the checksum field to immediately before the start of the Certificate Table entry in the
+		// optional header data directory.
+		// This is 60 bytes for PE32 format binaries, or 76 bytes for PE32+ format binaries.
+		sz := 60
+		if isPe32Plus {
+			sz = 76
+		}
+		b = make([]byte, sz)
+		if _, err := hr.Read(b); err != nil {
+			return nil, xerrors.Errorf("cannot read from checksum to certificate table data directory entry: %w", err)
+		}
+		h.Write(b)
+
+		// 6) Get the Attribute Certificate Table address and size from the Certificate Table entry.
+		certTable = &dd[certTableIndex]
+	}
+
+	// 7) Exclude the Certificate Table entry from the calculation and hash	everything from the end of the Certificate Table entry
+	// to the end of image header, including the Section Table. The Certificate Table entry is 8 bytes long.
+	if certTable != nil {
+		hr.Seek(8, io.SeekCurrent)
+	}
+	b, err = ioutil.ReadAll(hr)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot read remainder of headers and section table: %w", err)
+	}
+	h.Write(b)
+
+	// 8) Create a counter called sumOfBytesHashed, which is not part of the signature. Set this counter to the SizeOfHeaders field.
+	sumOfBytesHashed := sizeOfHeaders
+
+	// 9) Build a temporary table of pointers to all of the section headers in the image. Do not include any section headers in the
+	// table whose Size field is zero.
+	var sections []*pe.SectionHeader
+	for _, section := range p.Sections {
+		if section.Size == 0 {
+			continue
+		}
+		sections = append(sections, &section.SectionHeader)
+	}
+
+	// 10) Using the Offset field in the referenced SectionHeader structure as a key, arrange the table's elements in ascending order.
+	// In other words, sort the section headers in ascending order according to the disk-file offset of the sections.
+	sort.Slice(sections, func(i, j int) bool { return sections[i].Offset < sections[j].Offset })
+
+	for _, section := range sections {
+		// 11) Walk through the sorted table, load the corresponding section into memory, and hash the entire section. Use the
+		// Size field in the SectionHeader structure to determine the amount of data to hash.
+		sr := io.NewSectionReader(r, int64(section.Offset), int64(section.Size))
```

Comment:
should be check that section.Size is not unreasonably large? or chunk?

Reply:
I've add a helper to hash data 1 page at a time now